### ISSUE TITLE
Webhooks do not take array data as trigger.body

### ISSUE THREAD

**Issue by mickmcgrath13** [StackStorm/st2#3955]
I have a webhook:
```
curl -X POST https://127.0.0.1/api/v1/webhooks/my/webhook?st2-api-key=apikey -H "Content-Type: application/json" --data '[{"data":"val"}]' -k
```

And I get this error:
```
TypeError: type object argument after ** must be a mapping, not list (_exception_data={},_exception_class='TypeError',_exception_message='type object argument after ** must be a mapping, not list')
2018-01-22 20:03:20,288 140714391321168 INFO logging [-] fcc8f336-09c3-4989-a29b-caeb398f1e8c - 500 46 8.433ms (content_length=46,request_id='fcc8f336-09c3-4989-a29b-caeb398f1e8c',runtime=8.433,remote_addr='127.0.0.1',status=500,method='POST',path='/v1/webhooks/my/webhook')
```

..I'm going to assume this means that a service cannot send list data to a webhook, is that correct?
If so, this is new for 2.5 as it worked in 2.4.  We have services that send lists..

**Comment by Kami** [comment]
Thanks for reporting this.

I believe the issue is related to some frameworks not supporting arrays as root objects because of potential issues (https://haacked.com/archive/2009/06/25/json-hijacking.aspx/), but this only applies to responses and not requests such as in this case, so I think it's reasonable to make a change and also support arrays for ingress.

**Comment by Kami** [comment]
@humblearner I checked the code and as far as I know, this is not regression - it seems like that was always the case.

**Comment by Kami** [comment]
@mickmcgrath13 Can you please share your use case for an array with us? Could you update your code to wrap an array inside an object to avoid this limitation? Or is it an event from an external system and you can't do that?

I proposed a code change which would add support for arrays (https://github.com/StackStorm/st2/pull/3956), but @enykeev @lakshmi-kannan and I talked on Slack and we actually all agree it would be better to only support objects and not array - this would reduce code complexity and unnecessary edge cases.

**Comment by mickmcgrath13** [comment]
I was wrong about the previous version of stackstorm I mentioned.  The original version was `st2 2.2.1`.

The webhook is coming from a system for which we cannot modify the response.

I built a rule to handle the data sent from this service using stackstorm 2.2, and the rule worked as expected.  I then installed the same rule onto a 2.5 instance, and the same data broke the same rule (i.e. the only change I made was for the version of stackstorm).

Any ideas on how to get around this issue?

One option is to proxy through another server that does accept arrays.  It's not ideal, but it's not out of the realm of possibility for us.

**Comment by Kami** [comment]
Thanks for clarifying.

In this case (external system) the change does make sense and I will work on getting it merged tomorrow.

Until the change is merged and a new stable release with that change is out (you can try staging unstable packages if you feel adventurous after the change has been merged and new dev packages are built) you sadly don't have too many good options - as you mentioned, proxy which transforms data into a dict is one option.

Another option would be to write a sensor which receives a webhook and then dispatches a trigger into the system (we have multiple examples of such sensors in the exchange).

Or if you can wait a bit for the change to be merged - should be the easiest option and means you don't need to maintain and deal with the sensor, etc.

**Comment by Kami** [comment]
@mickmcgrath13 #3956 has been merged into master an will be available in v2.7.0dev packages soon.
